asdf_astropy/converters/transform/core.py

```diff
@@ -0,0 +1,206 @@
+
+        Returns
+        -------
+        dict
+            ASDF node.
+        """
+        pass # pragma: no cover
+
+    @abc.abstractmethod
+    def from_yaml_tree_transform(self, node, tag, ctx):
+        """
+        Convert an ASDF node into an instance of the appropriate
+        model class.  The implementing class need only instantiate
+        the model and set parameter values; common model properties
+        such as name and inverse will be handled by this base class.
+
+        Parameters
+        ----------
+        node : dict
+            The ASDF node to convert.
+        tag : str
+            The tag identifying the YAML type of the node.
+        ctx : asdf.asdf.SerializationContext
+            The context of the current serialization request.
+
+        Returns
+        -------
+        astropy.modeling.Model
+            The resulting model instance.
+        """
+        pass # pragma: no cover
+
+    def to_yaml_tree(self, model, tag, ctx):
+        from astropy.modeling.core import CompoundModel
+
+        node = self.to_yaml_tree_transform(model, tag, ctx)
+
+        if model.name is not None:
+            node["name"] = model.name
+
+        # Don't bother serializing analytic inverses provided
+        # by the model:
+        if getattr(model, "_user_inverse", None) is not None:
+            node["inverse"] = model._user_inverse
+
+        try:
+            bb = model.bounding_box
+        except NotImplementedError:
+            bb = None
+
+        if bb is not None:
+            if model.n_inputs == 1:
+                node["bounding_box"] = list(bb)
+            else:
+                node["bounding_box"] = [list(item) for item in bb]
+
+        # model / parameter constraints
+        if not isinstance(model, CompoundModel):
+            fixed_nondefaults = {k: f for k, f in model.fixed.items() if f}
+            if fixed_nondefaults:
+                node['fixed'] = fixed_nondefaults
+            bounds_nondefaults = {k: b for k, b in model.bounds.items() if any(b)}
+            if bounds_nondefaults:
+                node['bounds'] = bounds_nondefaults
+
+        return node
+
+    def from_yaml_tree(self, node, tag, ctx):
+        model = self.from_yaml_tree_transform(node, tag, ctx)
+
+        if 'name' in node:
+            model.name = node['name']
+
+        if 'bounding_box' in node:
+            model.bounding_box = node['bounding_box']
+
+        param_and_model_constraints = {}
+        for constraint in ['fixed', 'bounds']:
+            if constraint in node:
+                param_and_model_constraints[constraint] = node[constraint]
+
+        yield model
+
+        # TODO(eslavich): Is it necessary to initialize constraints
+        # after yielding?  Need a note explaining why, in any case.
+        model._initialize_constraints(param_and_model_constraints)
+        if 'inverse' in node:
+            model.inverse = node['inverse']
```

Comment:
Good question! Shouldn't the constraints be initialized on `__init__()` of the model?

Reply:
The model instance gets created by the concrete `Converter` classes, so I don't think we want to do it there since we'd have to duplicate this code everywhere.  I still can't think of why constraints have to be set before the yield, so I'll change this now and see if anything breaks when we test.